Python/Product/TestAdapter.Executor/Pytest/TestDiscovererPytest.cs

```diff
@@ -25,52 +41,94 @@ internal class TestDiscovererPytest : IPythonTestDiscoverer {
 
         public void DiscoverTests(IEnumerable<string> sources, IMessageLogger logger, ITestCaseDiscoverySink discoverySink) {
             _logger = logger;
+            string json = null;
 
-            var discoveryResults = new List<PytestDiscoveryResults>();
+            var workspaceText = _settings.IsWorkspace ? Strings.WorkspaceText : Strings.ProjectText;
+            LogInfo(Strings.PythonTestDiscovererStartedMessage.FormatUI(PythonConstants.PytestText, _settings.ProjectName, workspaceText,  _settings.DiscoveryWaitTimeInSeconds));
 
             try {
                 var env = InitializeEnvironment(sources, _settings);
                 var arguments = GetArguments(sources);
-
-                using (var outputStream = new MemoryStream())
-                using (var writer = new StreamWriter(outputStream, encoding: new UTF8Encoding(true), 4096, leaveOpen: true))
-                using (var proc = ProcessOutput.Run(
+                json = ExecuteProcess(
                     _settings.InterpreterPath,
+                    env,
                     arguments,
                     _settings.WorkingDirectory,
-                    env,
-                    visible: false,
-                    new StreamRedirector(writer)
-                )) {
+                    _settings.PathEnv,
+                    _settings.DiscoveryWaitTimeInSeconds
+                    );
+            } catch (TimeoutException) {
+                Error(Strings.PythonTestDiscovererTimeoutErrorMessage);
+                return;
+            }
 
-                    DebugInfo("cd " + _settings.WorkingDirectory);
-                    DebugInfo("set " + _settings.PathEnv + "=" + env[_settings.PathEnv]);
-                    DebugInfo(proc.Arguments);
+            List<PytestDiscoveryResults> results = null;
+            try {
+                results = JsonConvert.DeserializeObject<List<PytestDiscoveryResults>>(json);
+            } catch (InvalidOperationException ex) {
+                Error("Failed to parse: {0}".FormatInvariant(ex.Message));
+                Error(json);
+            } catch (JsonException ex) {
+                Error("Failed to parse: {0}".FormatInvariant(ex.Message));
+                Error(json);
+            }
 
-                    // If there's an error in the launcher script,
-                    // it will terminate without connecting back.
-                    WaitHandle.WaitAny(new WaitHandle[] { proc.WaitHandle });
+            CreateVsTests(results, discoverySink);
+        }
 
-                    outputStream.Flush();
-                    outputStream.Seek(0, SeekOrigin.Begin);
-                    var json = new StreamReader(outputStream).ReadToEnd();
 
+        private void CreateVsTests(IEnumerable<PytestDiscoveryResults> discoveryResults, ITestCaseDiscoverySink discoverySink) {
+            foreach (PytestDiscoveryResults result in discoveryResults ?? Enumerable.Empty<PytestDiscoveryResults>()) {
```

Comment:
We have an extension method that handles enumerating null as empty enumerable. Also use `var`.
```
foreach (var result in discoveryResults.MaybeEnumerate()) {
```